File: leetcode/js/find-the-prefix-common-array-of-two-arrays.js
/**
 * @param {number[]} A
 * @param {number[]} B
 * @return {number[]}
 */
var findThePrefixCommonArray = function (A, B) {
  const result = [];
  const set = new Set();
  let count = 0;

  for (let i = 0; i < A.length; i++) {
    if (set.has(A[i])) {
      count++;
    } else {
      set.add(A[i]);
    }

    if (set.has(B[i])) {
      count++;
    } else {
      set.add(B[i]);
    }

    result.push(count);
  }

  return result;
};

File: leetcode/rust/find-the-prefix-common-array-of-two-arrays.rs
impl Solution {
    pub fn find_the_prefix_common_array(a: Vec<i32>, b: Vec<i32>) -> Vec<i32> {
        let n = a.len();
        let mut num_count = vec![0; n + 1];
        let mut result = vec![0; n];

        let mut acc = 0;

        for i in 0..n {
            num_count[a[i] as usize] += 1;
            if num_count[a[i] as usize] == 2 {
                acc += 1;
            }

            num_count[b[i] as usize] += 1;
            if num_count[b[i] as usize] == 2 {
                acc += 1;
            }

            result[i] = acc;
        }

        result
    }
}
